src/kvstore/Common.h

```diff
@@ -39,6 +40,183 @@ class KVFilter {
+
+  std::string toString() const {
+    return addr.toString() + "," + std::to_string(static_cast<int>(status));
+  }
+
+  void fromString(const std::string& str) {
+    auto pos = str.find(",");
+    if (pos == std::string::npos) {
+      LOG(ERROR) << "Parse peer failed:" << str;
+      return;
+    }
+    addr = HostAddr::fromString(str.substr(0, pos));
+
+    int s = std::stoi(str.substr(pos + 1));
+    if (s >= static_cast<int>(Status::kMax) || s < 0) {
+      LOG(ERROR) << "Parse peer status failed:" << str;
+      return;
+    }
+
+    status = static_cast<Status>(s);
+  }
+
+  bool operator==(const Peer& rhs) const {
+    return addr == rhs.addr && status == rhs.status;
+  }
+
+  bool operator!=(const Peer& rhs) const {
+    return !(*this == rhs);
+  }
+
+  Peer& operator=(const Peer& rhs) {
+    addr = rhs.addr;
+    status = rhs.status;
+    return *this;
+  }
+
+  static Peer nullPeer() {
+    return Peer();
+  }
+};
+
+inline std::ostream& operator<<(std::ostream& os, const Peer& peer) {
+  return os << peer.toString();
+}
+
+/**
+ * @brief Peers for one partition, it should handle serializing and deserializing.
+ *
+ */
+struct Peers {
+  std::map<HostAddr, Peer> peers;
+
+  Peers() {}
+  explicit Peers(const std::vector<HostAddr>& addrs) {  // from normal peers
+    for (auto& addr : addrs) {
+      peers[addr] = Peer(addr, Peer::Status::kNormalPeer);
+    }
+  }
+  explicit Peers(const std::vector<Peer>& ps) {
+    for (auto& p : ps) {
+      peers[p.addr] = p;
+    }
+  }
+  explicit Peers(std::map<HostAddr, Peer> ps) : peers(std::move(ps)) {}
+
+  void addOrUpdate(const Peer& peer) {
+    peers[peer.addr] = peer;
+  }
+
+  bool get(const HostAddr& addr, Peer* peer) {
+    auto it = peers.find(addr);
+    if (it == peers.end()) {
+      return false;
+    }
+
+    if (peer != nullptr) {
+      *peer = it->second;
+    }
+    return true;
+  }
+
+  void remove(const HostAddr& addr) {
+    peers.erase(addr);
+  }
+
+  size_t size() {
+    return peers.size();
+  }
+
+  std::map<HostAddr, Peer> getPeers() {
+    return peers;
+  }
+
+  std::string toString() const {
+    std::stringstream os;
+    os << "version:1,"
+       << "count:" << peers.size() << "\n";
+    for (const auto& [_, p] : peers) {
+      os << p << "\n";
+    }
+    return os.str();
+  }
+
+  static std::pair<int, int> extractHeader(const std::string& header) {
+    auto pos = header.find(":");
+    if (pos == std::string::npos) {
+      LOG(ERROR) << "Parse part peers header error:" << header;
+      return {0, 0};
+    }
+    int version = std::stoi(header.substr(pos + 1));
+    pos = header.find(":", pos + 1);
+    if (pos == std::string::npos) {
+      LOG(ERROR) << "Parse part peers header error:" << header;
+      return {0, 0};
+    }
+    int count = std::stoi(header.substr(pos + 1));
+
+    return {version, count};
+  }
+
+  static Peers fromString(const std::string& str) {
+    Peers peers;
```

Comment:
You could try to use `folly::split` by `\n`.  BTW, the `count` in header is not actually necessary (could be calculated by counting `\n`)

Reply:
the count is redundant indeed, only to use check the count consistency.

Reply:
Ok, it is not a big deal. But you could try use `folly::split`, which will make code much more clear.

Reply:
have changed~